architecture as code intro blog post (#2671)
pulumi/docs: content/blog/architecture-as-code-intro/index.md

---
title: "Architecture as Code"
date: 2020-03-18T13:42:35-05:00

> meta_desc: " An overview of fnfrastructure architecture and implementations with Pulumi"

Reviewer: ```suggestion
meta_desc: " An overview of infrastructure architecture and implementations with Pulumi."
```

> meta_desc: " An overview of fnfrastructure architecture and implementations with Pulumi"

Reviewer 2: ```suggestion
meta_desc: " An overview of infrastructure architecture and implementations with Pulumi"
```

?

> meta_image: architecture.png

Reviewer: What are the objects in this image? I'm not sure I understand what the image is supposed to be communicating.

Author: Since the examples are in GCP, those are the logos for the GCP services. Amenable to change

authors:
    - sophia-parafina
tags:
    - virtual machines
    - Kubernetes
    - serverless
    - microservices
---

Abstraction is key to building resilient systems because it encapsulates behavior and decouples code, letting each component perform its function independently. The same principles apply to infrastructure, where we want to declare behavior or state and not implementation details. As an industry, we've moved away from monolithic applications to distributed systems such as serverless, microservices, Kubernetes, and virtual machine deployments. In this article, we'll take a closer look at the characteristics of these architectures and how Pulumi can abstract the components that comprise these systems.

<!--more-->

## Virtual Machines

> Virtual machines arose from the need to use expensive hardware more efficiently. VMs are more efficient because they share the host server, decreasing the number of physical servers needed to run applications. A group of hosts can also be aggregated so that it behaves as a single virtual host with large memory and processing capabilities. Virtual machines can emulate other environments that have specific hardware requirements or capabilities that the host machine does not have, thus reducing the need for specific hardware or operating systems. Finally, virtualization provides the host machine isolation from the guest virtual machines by limiting access to the host machine through an abstraction layer. Finally, virtualization has the added advantage of providing a fine degree of control over resources available to the virtual machine.

Reviewer 2: The last two sentences both start with "Finally,"

Author: fixed

> If virtual machines are part of your infrastructure, Pulumi lets you create virtual machines programmatically across cloud service providers. The Javascript example below illustrates how to deploy a virtual machine on Google Cloud Platform. The script spins up a Debian based instance and creates a network with an open port 22 to let you configure and manage the machine as needed.

Reviewer 2: ```suggestion
If virtual machines are part of your infrastructure, Pulumi lets you create virtual machines programmatically across cloud service providers. The JavaScript example below illustrates how to deploy a virtual machine on Google Cloud Platform. The script spins up a Debian based instance and creates a network with an open port 22 to let you configure and manage the machine as needed.
```

```js
const gcp = require("@pulumi/gcp");

> const fs = require(‘fs’);

Reviewer: ```suggestion
const fs = require("fs");
```

> const fs = require(‘fs’);

Reviewer 2: ```suggestion
const fs = require("fs");
```

// Create a network
const network = new gcp.compute.Network("network");
const computeFirewall = new gcp.compute.Firewall("firewall", {
    network: network.id,
    allows: [{
        protocol: "tcp",

>         ports: [ "22",”80” ],

Reviewer: ```suggestion
        ports: [ "22", "80" ],
```

    }],
});

// Read the configuration script

> const startupScript = fs.readFileSync(‘config.sh’, ‘utf-8’)

Reviewer: This doesn't work when copied because of the smart quotes:

```suggestion
const startupScript = fs.readFileSync("config.sh", "utf-8")
```

It's good to try coping these yourself with the Copy button, just be sure Hugo didn't mangle anything. 😄

Reviewer: It's good practice just to try coping these with the copy button, to be sure Hugo didn't mangle anything. 😄

> const startupScript = fs.readFileSync(‘config.sh’, ‘utf-8’)

Reviewer 2: ```suggestion
const startupScript = fs.readFileSync("config.sh", "utf-8")
```

> // Create a Virtual Machine Instance

Reviewer: ```suggestion
// Create a Virtual Machine Instance.
```

const computeInstance = new gcp.compute.Instance("instance", {
    machineType: "f1-micro",
    zone: "us-central1-a",
    metadataStartupScript: startupScript,
    bootDisk: { initializeParams: { image: "debian-cloud/debian-9" } },
    networkInterfaces: [{
        network: network.id,

>         // accessConfigus must includ a single empty config to request an ephemeral IP

Reviewer: ```suggestion
        // accessConfigs must include a single empty config to request an ephemeral IP.
```

        accessConfigs: [{}],
    }],
});

> // Export the name and IP address of the Instance

Reviewer: Nits, but I think we generally try to have comments be complete sentences. 

```suggestion
// Export the name and IP address of the Instance.
```

exports.instanceName = computeInstance.name;
exports.instanceIP = computeInstance.networkInterfaces.apply(ni => ni[0].accessConfigs[0].natIp);
```

> This is the configuration script.

Reviewer: ```suggestion
This is the configuration script, `config.sh`, referenced in the code:
```

```bash
#!/bin/bash
echo "Hello, World!" > index.html
nohup python -m SimpleHTTPServer 80 &
```

When the virtual machine is created, it reads the configuration file that creates a webserver. This a basic example, but it shows how to configure virtual machines programmatically, which is useful for scaling horizontally.

## Serverless

> Serverless is a computing architecture characterized by server-side logic running in stateless containers that are invoked by events. They are typically ephemeral (often are available for only one call), and managed by third-party cloud providers. Serverless is also called Functions as a Service, or `FaaS`, and well-known implementations include AWS Lambda and Fargate, Azure Functions, and Google Cloud Run and Cloud Functions.

Reviewer: It's not code, so doesn't need to be wrapped in a `<code>` element:

```suggestion
Serverless is a computing architecture characterized by server-side logic running in stateless containers that are invoked by events. They are typically ephemeral (often are available for only one call), and managed by third-party cloud providers. Serverless is also called Functions as a Service, or FaaS, and well-known implementations include AWS Lambda and Fargate, Azure Functions, and Google Cloud Run and Cloud Functions.
```

> A significant benefit of serverless is that it's a polyglot platform that allows developers to choose languages optimized for a task. For example, scripting languages like Javascript or Python may be more responsive than a language such as Java. Serverless functions can support both synchronous and asynchronous calls. There are use cases that require an immediate response such as processing a video stream and instances where returning the result immediately is not critical as in an ETL batch job.

Reviewer 2: Javascript => JavaScript (applies throughout)

```suggestion
A significant benefit of serverless is that it's a polyglot platform that allows developers to choose languages optimized for a task. For example, scripting languages like JavaScript or Python may be more responsive than a language such as Java. Serverless functions can support both synchronous and asynchronous calls. There are use cases that require an immediate response such as processing a video stream and instances where returning the result immediately is not critical as in an ETL batch job.
```

Reviewer 2: > For example, scripting languages like Javascript or Python may be more responsive than a language such as Java.

Really? Just because startup time?

A key part of serverless architecture is an API Gateway that provides logical routes for mapping standard HTTP operations such as GET, PUT, POST, and DELETE to functions. The gateway makes development easier since these are standard and well-known interfaces. In addition to an API Gateway, the serverless platform should provide REST endpoints that allow you to manage the deploy with a CLI, portal, or an automation script. Finally, serverless architecture is extensible and supports integration with event sources and resources from the cloud provider through webhooks and other mechanisms.

> The following example illustrates how to deploy two functions, one written in Python and the other written in Go. You can find the [full example](https://github.com/pulumi/examples/tree/master/gcp-ts-serverless-raw) in the Pulumi Github repository.

Reviewer 2: ```suggestion
The following example illustrates how to deploy two functions using TypeScript, one written in Python and the other written in Go. You can find the [full example](https://github.com/pulumi/examples/tree/master/gcp-ts-serverless-raw) in the Pulumi GitHub repository.
```

const bucket = new gcp.storage.Bucket("bucket");

// Google Cloud Function in Python

const bucketObjectPython = new gcp.storage.BucketObject("python-zip", {
    bucket: bucket.name,
    source: new asset.AssetArchive({
        ".": new asset.FileArchive("./pythonfunc"),
    }),
});

const functionPython = new gcp.cloudfunctions.Function("python-func", {
    sourceArchiveBucket: bucket.name,
    runtime: "python37",
    sourceArchiveObject: bucketObjectPython.name,
    entryPoint: "handler",
    triggerHttp: true,
    availableMemoryMb: 128,
});

export const pythonEndpoint = functionPython.httpsTriggerUrl;

// Google Cloud Function in Go

const bucketObjectGo = new gcp.storage.BucketObject("go-zip", {
    bucket: bucket.name,
    source: new asset.AssetArchive({
        ".": new asset.FileArchive("./gofunc"),
    }),
});

const functionGo = new gcp.cloudfunctions.Function("go-func", {
    sourceArchiveBucket: bucket.name,
    runtime: "go111",
    sourceArchiveObject: bucketObjectGo.name,
    entryPoint: "Handler",
    triggerHttp: true,
    availableMemoryMb: 128,
});

export const goEndpoint = functionGo.httpsTriggerUrl;
```

## Kubernetes

> Kubernetes is a container orchestration system for deploying, scaling, and managing your application. Applications use multiple containers that contain code and dependencies needed for the code to execute. Kubernetes runs on a cluster of nodes and containers are deployed in pods. Kubernetes manages the pods by replacing containers that fail and killing containers that don't respond. It can also scale the number of containers as needed and load balance the network traffic. These are some of the features that Kubernetes provides:

Reviewer: Is "killing" the right word? If it's not a Kubernetes-specific term, maybe something less aggressive, like destroying or disposing of.

Author: substituted with remove

                this.service.status.apply(status => status.loadBalancer.ingress[0].ip);
        }
    }
}

export interface ServiceDeploymentArgs {
    image: string;
    resources?: k8stypes.core.v1.ResourceRequirements;
    replicas?: number;
    ports?: number[];
    allocateIpAddress?: boolean;
    isMinikube?: boolean;
}
```

The main program deploys an instance of Redis and the guestbook application using containers pulled from the Google Container Registry.

```ts
import * as pulumi from "@pulumi/pulumi";
import * as k8sjs from "./k8sjs";

const config = new pulumi.Config();

const redisMaster = new k8sjs.ServiceDeployment("redis-master", {
    image: "k8s.gcr.io/redis:e2e",
    ports: [6379],
});

const redisReplica = new k8sjs.ServiceDeployment("redis-replica", {
    image: "gcr.io/google_samples/gb-redisslave:v1",
    ports: [6379],
});

const frontend = new k8sjs.ServiceDeployment("frontend", {
    replicas: 3,
    image: "gcr.io/google-samples/gb-frontend:v4",
    ports: [80],
    allocateIpAddress: true,
    isMinikube: config.getBoolean("isMinikube"),
});

export let frontendIp = frontend.ipAddress;
```

> The [original example](https://github.com/kubernetes/examples/tree/master/guestbook) defines the Service and Deployment in YAML. In the example above, the YAML is abstracted in a reusable component written in a modern programming language. The [complete example](https://github.com/pulumi/examples/tree/master/kubernetes-ts-guestbook) is available on Github. As you can see, the ability to reuse components gets us closer to implementing architecture as code.

Reviewer: ```suggestion
The [original example](https://github.com/kubernetes/examples/tree/master/guestbook) defines the Service and Deployment in YAML. In the example above, the YAML is abstracted in a reusable component written in a modern programming language. The [complete example](https://github.com/pulumi/examples/tree/master/kubernetes-ts-guestbook) is available on GitHub. As you can see, the ability to reuse components gets us closer to implementing architecture as code.
```

> The [original example](https://github.com/kubernetes/examples/tree/master/guestbook) defines the Service and Deployment in YAML. In the example above, the YAML is abstracted in a reusable component written in a modern programming language. The [complete example](https://github.com/pulumi/examples/tree/master/kubernetes-ts-guestbook) is available on Github. As you can see, the ability to reuse components gets us closer to implementing architecture as code.

Reviewer 2: Github => GitHub (applies throughout)

```suggestion
The [original example](https://github.com/kubernetes/examples/tree/master/guestbook) defines the Service and Deployment in YAML. In the example above, the YAML is abstracted in a reusable component written in a modern programming language. The [complete example](https://github.com/pulumi/examples/tree/master/kubernetes-ts-guestbook) is available on GitHub. As you can see, the ability to reuse components gets us closer to implementing architecture as code.
```

## Microservices

> Microservices are based on the idea that components encapsulate a business capability. Services are independently deployable and communicate via web service requests, which has the advantage of redeploying only one or selected services and leaving the remaining components up and running. The [12-Factor app](https://12factor.net/) is the canonical pattern for microservice, which is summarized by:

Reviewer: ```suggestion
Microservices are based on the idea that components encapsulate a business capability. Services are independently deployable and communicate via web service requests, which has the advantage of redeploying only one or selected services and leaving the remaining components up and running. The [12-Factor app](https://12factor.net/) is the canonical pattern for microservices, which is summarized by:
```

- Use declarative methods to configure and deploy applications.
- Enforce maximum portability between environments by establishing a clean contract with the underlying operating system.
- Deploy on modern cloud platforms that diminish the need for servers and the requisite system administration.
- Establish continuous deployment by keeping development, testing, and production stages similar.
- Use platforms with the ability to scale without significant changes to architecture, development, or tooling.

> In addition to Kubernetes, there are other platforms for deploying container-based applications. AWS Elastic Container Service, Google Cloud Run, and Azure Container Service provide alternatives to Kubernetes for container orchestration. The following code snippet demonstrates deploying a container with a Ruby application in Google Cloud Run. Note that the image is built using a local Docker engine and pushed to the Google Cloud Repository. The code shows that you can also set memory limits and concurrency for the container. The [full example](https://github.com/pulumi/examples/tree/master/gcp-ts-cloudrun) is available on Github.

Reviewer: ```suggestion
In addition to Kubernetes, there are other platforms for deploying container-based applications. AWS Elastic Container Service, Google Cloud Run, and Azure Container Service provide alternatives to Kubernetes for container orchestration. The following code snippet demonstrates deploying a container with a Ruby application in Google Cloud Run. Note that the image is built using a local Docker engine and pushed to the Google Cloud Repository. The code shows that you can also set memory limits and concurrency for the container. The [full example](https://github.com/pulumi/examples/tree/master/gcp-ts-cloudrun) is available on GitHub.
```
